symengine/tests/basic/test_polynomial.cpp

```diff
@@ -302,3 +304,197 @@ TEST_CASE("Univariate Polynomial expand", "[UnivariatePolynomial][expand]")
+
+    RCP<const MultivariateIntPolynomial> P = MultivariateIntPolynomial::from_dict({x,y}, { {{1,2}, 1_z}, {{1,1}, 2_z}, {{0,1}, 2_z}, {{1,0}, 3_z} , {{0,0}, 0_z} });
+    REQUIRE(P->__str__() == "x y**2 + 2*x y + 3*x + 2*y");
+
+    RCP<const MultivariateIntPolynomial> P2 = MultivariateIntPolynomial::from_dict({x,y}, { {{0,0}, 0_z} });
+    REQUIRE(P2->__str__() == "0");
+
+}
+
+TEST_CASE("Testing MultivariateIntPolynomial::__hash__() and compare", "[MultivariateIntPolynomial]"){
+    RCP<const Symbol> a = symbol("a");
+    RCP<const Symbol> b = symbol("b");
+    RCP<const Symbol> x = symbol("x");
+    RCP<const Symbol> y = symbol("y");
+    RCP<const MultivariateIntPolynomial> p1 = MultivariateIntPolynomial::from_dict({x,y}, {{{2,0},1_z}, {{1,1},1_z}, {{0,2},1_z}  });
+    RCP<const MultivariateIntPolynomial> p2 = MultivariateIntPolynomial::from_dict({x,y}, {{{2,0},1_z}, {{1,1},-1_z}, {{0,2},1_z}  });
+    RCP<const MultivariateIntPolynomial> p3 = MultivariateIntPolynomial::from_dict({x,y}, {{{2,0},2_z}, {{0,2},2_z}  });
+    RCP<const MultivariateIntPolynomial> p4 = MultivariateIntPolynomial::from_dict({a,b}, {{{2,0},2_z}, {{0,2},2_z}  });
+
+    //Only requre that the same polynomial hash to the same value and that different polynomials hash to different values
+    //Don't want to require a polynomial to have a particular hash in case someone comes up with a better hash function
+    REQUIRE(p3->__hash__() == add_mult_poly(*p1, *p2)->__hash__());
+    REQUIRE(p1->__hash__() != p2->__hash__());
+    REQUIRE(p3->__hash__() != p4->__hash__());
+
+    //Same for compare.
+    REQUIRE(0 == p3->compare(*add_mult_poly(*p1, *p2)));
+    REQUIRE(0 != p1->compare(*p2));
+    REQUIRE(0 != p3->compare(*p4));
+}
+
+TEST_CASE("Testing MultivariateIntPolynomial::__eq__(const Basic &o)", "[MultivariateIntPolynomial]")
+{
+    RCP<const Symbol> x = symbol("x");
+    RCP<const Symbol> y = symbol("y");
+    RCP<const MultivariateIntPolynomial> p1 = MultivariateIntPolynomial::from_dict({x,y}, {{{2,0},1_z}, {{1,1},1_z}, {{0,2},1_z}  });
+    RCP<const MultivariateIntPolynomial> p2 = MultivariateIntPolynomial::from_dict({x,y}, {{{2,0},1_z}, {{1,1},-1_z}, {{0,2},1_z}  });
+    RCP<const MultivariateIntPolynomial> p3 = MultivariateIntPolynomial::from_dict({x,y}, {{{2,0},2_z}, {{0,2},2_z}  });
+
+    REQUIRE(p1->__eq__(*p1));
+    REQUIRE(!(p2->__eq__(*p1)));
+    REQUIRE(p3->__eq__( *add_mult_poly(*p1, *p2) ));
+
+}
+
+TEST_CASE("Testing MultivariateIntPolynomial::eval((std::map<RCP<const Symbol>, integer_class, RCPSymbolCompare> &vals)", "[MultivariateIntPolynomial]")
+{
+    RCP<const Symbol> x = symbol("x");
+    RCP<const Symbol> y = symbol("y");
+    RCP<const Symbol> z = symbol("z");
+    RCP<const MultivariateIntPolynomial> p = MultivariateIntPolynomial::from_dict( {x,y,z}, { {{2,0,0},1_z}, {{0,2,0},2_z}, {{0,0,2},3_z}, {{1,1,1},4_z}, {{1,1,0},1_z}, {{0,1,1},2_z}, {{1,0,0},1_z}, {{0,1,0},2_z}, {{0,0,1}, 3_z} , {{0,0,0},5_z} });
```

Comment:
Fewer characters per line help in code review. I follow 80 chars per line; you could choose your own limit.